### PR TITLE
Network/refactoring/select/15 network refactoring select

### LINKED ISSUE #15: Network - Refactoring - Select
Implement the select process in the network library.

Create a function named : server_fill_fdsets who do the following :
- Clean the fdsets
- Fill the readfds according to the data stocked in the  client structure. If "pending_command" != true, add in the readfd
- Adding the server in the readfds
- Fill the writeFds according to the data stocked in the client structure. If "pending_writing" == true, add in the writefd

Create a function named : server_wait who do the following :
- Launch the select function with the Fdsets stocked locally, hold a potential timer and erros.

Create a function named : server_manage_fd_update :
- Read the readfds, if the server is find hold the new_connection_demand process
- Read the readfds, if a client is find hold the read_client process
- Read the writefds, if a client is find hold the write_client process

### PR DESCRIPTION
**Pull request by Cavonstavant**
Closes #15 #16 #17 #18

### CODE REVIEW

**Review by osvegn — CHANGES_REQUESTED** [review]

File: server/net_utils/display_utils.c [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    CIRCLEQ_FOREACH(tmp, peers_head, peers)
+        printf(" - %s:%hu on socket: %d\n",
+            inet_ntoa(tmp->peer_addr.sin_addr),
+            ntohs(tmp->peer_addr.sin_port), tmp->sock_fd);
+    printf("/*----------------------*/\n");
+}
+
+void display_start_message(struct sockaddr_in addr)
+{
+    char *host_repr = malloc(sizeof(char) * 16);
```
Comment (LucasTesnier) [inline server/net_utils/display_utils.c:26]: Malloc not verified

File: server/net_utils/tcp_server.c [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+tcp_server_t *create_tcp_server(long port)
+{
+    tcp_server_t *server = malloc(sizeof(tcp_server_t));
+
+    if (!server)
+        TEAMS_LOG("malloc");
+    server->port = htons(port);
+    bind_and_listen_on_socket(server, port);
+    FD_ZERO(&server->read_fds);
```
Comment (LucasTesnier) [inline server/net_utils/tcp_server.c:47]: Why modify the FDSETS here?

**Review by LucasTesnier — CHANGES_REQUESTED** [review]
NTR!

File: server/net_utils/tcp_server.c (hunk 2 above)
Comment (Cavonstavant) [reply server/net_utils/tcp_server.c:47]: it's like a calloc but with FD sets @LucasTesnier

**Review by osvegn — APPROVED** [review]
LGTM

**Review by osvegn — CHANGES_REQUESTED** [review]
PR documentation

File: server/net_utils/auth/user.c [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+    if (!user)
+        return (NULL);
+    user->username = (username ? strdup(username) : NULL);
+    user->password = (password ? strdup(password) : NULL);
+    user->logged_in = true;
+    return (user);
+}
+
+enum authentification_e authenticate_user(struct users_head users,
```
Comment (LucasTesnier) [inline server/net_utils/auth/user.c:26]: Typedef ?

File: server/net_utils/fd_sets_manage/fd_set_manage.h [hunk 4]
```diff
@@ -0,0 +1,34 @@
+/*
+** EPITECH PROJECT, 2022
+** mes_equipes
+** File description:
+** fd_set_manage
+*/
+
+#ifndef FD_SET_MANAGE_H
+#define FD_SET_MANAGE_H
```
Comment (LucasTesnier) [inline server/net_utils/fd_sets_manage/fd_set_manage.h:9]: Indentation not Epitech Copliant

File: server/net_utils/fd_sets_manage/fd_set_manage.h [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+    #include "net_utils.h"
+
+/// \brief Accepts a new client
```
Comment (LucasTesnier) [inline server/net_utils/fd_sets_manage/fd_set_manage.h:13]: Missing documentation

File: server/net_utils/fd_sets_manage/fd_sets_manage.c [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+** mes_equipes
+** File description:
+** fd_sets_manage
+*/
+
+/// \file server/net_utils/fd_set_manage.c
+
+#include "net_utils.h"
+#include "fd_set_manage.h"
+#include <sys/time.h>
```
Comment (LucasTesnier) [inline server/net_utils/fd_sets_manage/fd_sets_manage.c:12]: Why ?

File: server/net_utils/fd_sets_manage/server_accept_new_client.c [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+    socklen_t client_addr_size = sizeof(client_addr);
+
+    client_fd = accept(srv->sock_fd,
+        (struct sockaddr *)&client_addr,
+        &client_addr_size);
+    if (client_fd < 0)
+        TEAMS_LOG("accept");
+    new_client = new_peer(client_fd, client_addr);
+    if (new_client == NULL)
+        TEAMS_LOG("peer_create");
```
Comment (LucasTesnier) [inline server/net_utils/fd_sets_manage/server_accept_new_client.c:24]: Return ?

File: server/net_utils/fd_sets_manage/server_accept_new_client.c [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+    new = accept_client(srv);
+    if (new == NULL){
+        TEAMS_LOG("Internal Error: could not accept new client.\n");
+        return (false);
+    }
+    CIRCLEQ_INSERT_HEAD(&srv->peers_head, new, peers);
+    return (true);
+}
+
+bool server_accept_new_client(tcp_server_t *srv)
```
Comment (LucasTesnier) [inline server/net_utils/fd_sets_manage/server_accept_new_client.c:40]: WHYYYYYY ?

File: server/net_utils/fd_sets_manage/server_read_client.c [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+        return (NULL);
+    if (!peer->pending_read)
+        return (NULL);
+    peer->pending_read = false;
+    return (strdup(peer->output_buffer));
+}
+
+bool server_read_client(tcp_server_t *srv, peer_t *tmp)
+{
+    if (!tmp && !srv)
```
Comment (LucasTesnier) [inline server/net_utils/fd_sets_manage/server_read_client.c:22]: ||

File: server/net_utils/fd_sets_manage/server_read_client.c [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+        return (false);
+    ssize_t read_size;
```
Comment (LucasTesnier) [inline server/net_utils/fd_sets_manage/server_read_client.c:24]: Init in body function

File: server/net_utils/fd_sets_manage/server_accept_new_client.c [hunk 11]
```diff
@@ -0,0 +1,45 @@
+/*
+** EPITECH PROJECT, 2022
+** mes_equipes
+** File description:
+** server_accept_new_client
+*/
+
+#include "net_utils.h"
+
+static peer_t *accept_client(tcp_server_t *srv)
```
Comment (LucasTesnier) [inline server/net_utils/fd_sets_manage/server_accept_new_client.c:10]: Missing doc

File: server/net_utils/fd_sets_manage/server_accept_new_client.c [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+    if (client_fd < 0)
+        TEAMS_LOG("accept");
+    new_client = new_peer(client_fd, client_addr);
+    if (new_client == NULL)
+        TEAMS_LOG("peer_create");
+    FD_SET(client_fd, &srv->read_fds);
+    return (new_client);
+}
+
+static bool manage_new_conn(tcp_server_t *srv, peer_t *new)
```
Comment (LucasTesnier) [inline server/net_utils/fd_sets_manage/server_accept_new_client.c:29]: Missing doc

**Review by LucasTesnier — CHANGES_REQUESTED** [review]
Some changes to do

**Review by LucasTesnier — APPROVED** [review]
LGTM!

File: server/net_utils/fd_sets_manage/server_read_client.c [hunk 13]
```diff
@@ -0,0 +1,37 @@
+/*
+** EPITECH PROJECT, 2022
+** mes_equipes
+** File description:
+** server_read_client
+*/
+
```
Comment (osvegn) [inline server/net_utils/fd_sets_manage/server_read_client.c:7]: missing \file

File: server/net_utils/fd_sets_manage/server_read_client.c [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+#include "net_utils.h"
+
+char *fetch_message(peer_t *peer)
+{
+    if (!peer)
+        return (NULL);
+    if (!peer->pending_read)
+        return (NULL);
+    peer->pending_read = false;
+    return (strdup(peer->output_buffer));
```
[the remark this answers is not in the document]
Comment (osvegn) [reply server/net_utils/fd_sets_manage/server_read_client.c:17]: or not

[the remark this answers is not in the document]
Comment (osvegn) [reply server/net_utils/fd_sets_manage/server_read_client.c:17]: i don't know

File: server/net_utils/fd_sets_manage/server_write_client.c [hunk 15]
```diff
@@ -0,0 +1,43 @@
+/*
+** EPITECH PROJECT, 2022
+** mes_equipes
+** File description:
+** server_write_client
+*/
+
```
Comment (osvegn) [inline server/net_utils/fd_sets_manage/server_write_client.c:7]: missing \file

**Review by osvegn — CHANGES_REQUESTED** [review]
Request minor changes, but LGTM

File: server/net_utils/fd_sets_manage/server_read_client.c (hunk 14 above)
[the remark this answers is not in the document]
Comment (Cavonstavant) [reply server/net_utils/fd_sets_manage/server_read_client.c:17]: you're right i forgot

**Review by osvegn — APPROVED** [review]
LGTM

**Pull request merged**